WIP: make OB a generic cmake project

yetist opened the pull request:
目前的 OB 存在以下一些问题，不利于移植到新发行版和新架构：

1. 仅支持 openssl-1.0，不支持 openssl-1.1
2. 限制bison为2.4.1，无法使用高版本
3. 限制gcc-5.2.0，llvm-11.0.1等工具链版本，无法使用高版本
4. 静态编译，对发行版不友好，并非所有发行版都会提供静态依赖库。
5. `deps/logmessage` 子模块在每次编译时都会从 github 同步，对国内用户不友好。
6. 目前的项目构建方式太另类，无法直接使用 cmake 命令构建，且仅支持非常有限的发行版。
...

如果以上问题可以得到有效解决，将会使 OB 成为一个与发行版无关的 **通用 cmake 项目**，可以直接使用发行版自身的开发工具和依赖库来完成构建，有利于扩大 OB 的使用范围。

要完成以上目标，不仅要修改 CMake 工程，还要改代码，并在多个发行版上进行测试验证，工作量不小，希望在此可以进行讨论、测试，并推动 OB 向这个方向前进。

本 PR 对此做了初步尝试，目前可在 Archlinux 上以 Debug 方式编译（Release 方式编译会出错，见build2.sh）。

yetist committed:
Fix syntax to use higher version bison

yetist committed:
Use the toolchain provided by the distro

yetist committed:
Remove -Werror to use a later version of the toolchain

yetist committed:
update gitsubmodule

yetist committed:
use pkg-config to check openssl module

yetist committed:
Remove non-configurable build parameters

yetist committed:
fix build error

yetist committed:
Fix mariadb header paths when using pkgconfig

yetist committed:
fix build error

yetist committed:
fix build error

yetist committed:
Regenerate ob_error.{h,cpp} at build time

yetist committed:
link shared libraries on the system

yetist committed:
[cmake] Don't sync git submodules on every build

yetist committed:
use pkgconfig to check RapidJSON and isa-l module

yetist committed:
Add support for openssl-1.1

yetist commented:
@whhe @LINxiansheng 

使用 `cmake -DCMAKE_BUILD_TYPE=Debug ...` 可正常编译出 `observer`；而使用`cmake -DCMAKE_BUILD_TYPE=Release ...` 编译时，在最后链接时会报错如下：
```
[100%] Linking CXX executable observer                                           
cd /home/yetist/checkout/04-loongarch/oceanbase/build/src/observer && /usr/bin/cmake -E cmake_link_script CMakeFiles/observer.dir/link.txt --verbose=1
/usr/bin/g++ -fdebug-prefix-map=/home/yetist/checkout/04-loongarch/oceanbase=. -fdiagnostics-color -O3 -DNDEBUG -Wl,--build-id=uuid -z noexecstack CMakeFiles/ob_main.dir/main.cpp.o -o observer  -Wl,-rpath,/home/yetist/checkout/04-loongarch/oceanbase/build/src/observer:/home/yetist/checkout/04-loongarch/oceanbase/build/deps/oblib/src:/home/yetist/checkout/04-loongarch/oceanbase/build/deps/easy/src liboceanbase_static.so -lgcc -lstdc++ ../../deps/oblib/src/lib/libmalloc_hook.a -L/home/yetist/checkout/04-loongarch/oceanbase/deps/3rd/usr/local/oceanbase/deps/devel/lib/mariadb -lmariadb -laio ../../deps/oblib/src/liboblib.so ../../deps/easy/src/libeasy.so -lisal -lunwind -ldl -lrt -lcrypto -lcurl -L/home/yetist/checkout/04-loongarch/oceanbase/deps/3rd/usr/local/oceanbase/deps/devel/lib -L/usr/lib -lssl -lcrypto -lpthread -lm 
/usr/bin/ld: liboceanbase_static.so: undefined reference to `oceanbase::sql::ObQueryRange::get_param_value(oceanbase::common::ObObj&, oceanbase::common::Ob2DArray<oceanbase::common::ObObjParam, 2096128, oceanbase::common::ObWrapperAllocator, false, oceanbase::common::ObSEArray<oceanbase::common::ObObjParam*, 64l, oceanbase::common::ObWrapperAllocator, false> > const&) const'
/usr/bin/ld: liboceanbase_static.so: undefined reference to `oceanbase::obrpc::ObLogicPartitionServiceRpcP<(oceanbase::obrpc::ObRpcPacketCode)1134>::~ObLogicPartitionServiceRpcP()'
/usr/bin/ld: liboceanbase_static.so: undefined reference to `oceanbase::obrpc::ObLogicPartitionServiceRpcP<(oceanbase::obrpc::ObRpcPacketCode)1135>::~ObLogicPartitionServiceRpcP()'
collect2: 错误：ld 返回 1                                                        
make[2]: *** [src/observer/CMakeFiles/observer.dir/build.make:90：src/observer/observer] 错误 1
```

请帮忙看看是代码存在问题还是工程存在问题？谢谢

yetist commented:
> hi, 非常感谢您的贡献，在更新构建工具链的问题上OceanBase一直持谨慎的态度，我们的研发团队认为，当前白名单工具链的方式还是需要保留（比如静态链接依赖库），一方面，我们需要保证现在已经上线投产的开源用户的稳定性，另一方面，我们现在的人力只能支持白名单工具链下产物的运维。我们也欢迎您在在不影响原有设计的基础上，扩展OS兼容表；

PR 标题增加了 `WIP:`，就是希望可以在充分讨论、沟通的基础上，边测试验证边完善，尽量做到在不影响现有的静态编译方式基础上，增加动态编译、支持高版本工具链和库、支持更多的发行版等特性，将 OB 改造成通用 cmake 项目，在这个过程中产生的一些合理且共性的修改，将创建单独的 PR 发起合并。

这个目标的实现需要大家一起参与，因为涉及到代码修改、工程改造，各个发行版测试反馈等大量工作要做，当然，如果没人参与，甚至 OB 社区不赞同这种做法的话，那就没必要继续浪费时间在这个上面了。

xen0n commented:
目前这种锚定依赖版本的做法，实质上排除了客户在（阿里数据库服务器不采用的）新兴架构上自愿运行 OceanBase 的可能性。一些国内外近年才出现的新兴架构，如 RISC-V 或 LoongArch，并不被锚定的老旧依赖版本所支持，但实际上作为技术的上游提供者，相信你们也并不希望这些人被排除在目标用户群体之外。

一种解决方法是引入**分层支持** (tiered support) 的概念。以 [Rust 语言的平台支持策略](https://doc.rust-lang.org/rustc/platform-support.html)为参考，可以考虑以下的划分：

* Tier-1：将获得完整的 CI 覆盖、测试覆盖、技术支持；包括目前你们所“官方支持”的所有架构、操作系统组合。
* Tier-2：代码有支持，CI 依赖相应维护者但不保证，没有技术支持；包括 RISC-V、LoongArch 等架构。
* Tier-3：所有其他场景，对功能不作任何保证（No Warranty）。

longdafeng commented:
非常感谢你们的pr, 我们也特别欢迎更多的pr.  

我们的用户有几百家金融用户, 而且我们是一个oltp的数据库, 承载着无数金融核心系统的稳定和安全, 我们不期望因为一些其他的缘故, 而导致莫名其妙的错误. 

另外一个事实, 就是我所知道的商业数据库, 都是一个稳定的工具链, 不是让数据库在任何环境下都能随意的编译.  另外, 我们曾经发生过因为升级工具链而发生过故障, 所以这个事情, 我们是有教训的. 

可能有几个原则, 我需要稍微提一下:
1. 工具链, 我们内部有专门的编译团队来评估工具链,  如果现在RISC-V 或 LoongArch 不能提供我们相应的版本, 麻烦你们说一下你们能提供什么版本, 我们可以做一些评估, 是否可以升级到对应的版本. 
2. deps/logmessage 子模块  这个问题, 已经在fix了, 应该下个版本就ok了

xen0n commented:
感谢及时回复！

> 我们的用户有几百家金融用户, 而且我们是一个oltp的数据库, 承载着无数金融核心系统的稳定和安全, 我们不期望因为一些其他的缘故, 而导致莫名其妙的错误.
>
> 另外一个事实, 就是我所知道的商业数据库, 都是一个稳定的工具链, 不是让数据库在任何环境下都能随意的编译. 另外, 我们曾经发生过因为升级工具链而发生过故障, 所以这个事情, 我们是有教训的.

理解这个需求，不过可能这场景和新兴架构支持并不冲突——从商务、技术、法律视角看，只要明确界定何种配置（例如，必须采用官方发布的二进制，必须满足一定的系统基础组件版本范围之类）才适用稳定性、安全性的承诺，那么其他人自己折腾应该是自愿、自由的，这也和开源精神一致。

> 工具链, 我们内部有专门的编译团队来评估工具链, 如果现在RISC-V 或 LoongArch 不能提供我们相应的版本, 麻烦你们说一下你们能提供什么版本, 我们可以做一些评估, 是否可以升级到对应的版本.

不止是工具链，其他依赖可能也涉及，但 RV 或者 LoongArch 的场景下基本直接怼到最新才有可能支持。（FYI，LoongArch gcc port 昨天才进上游，glibc 截至目前还没进，但 glibc 貌似是你们隐式依赖（我没仔细看，但没见到有完整的交叉编译工具链结构，所以假定如此），所以也不太影响。）这里出于方便起见，可能还是得支持使用系统组件或者自定义组件路径。这些新架构无一例外经常有人在做实验，用外界控制的依赖版本构建是刚需。

> 整体上, 我们还是期望我们的ob 能跑在LoongArch上, 支持更多的os和芯片, 尤其是我们自己的, 现在就是看我们怎么来解决这些问题而又不带来风险.

如前所述，我觉得这个主要还是文档（政策）、流程方面需要考虑。支持多个构建方式本身就很正常，被 Linux 发行版打包的软件往往都有两套构建系统，上游自己发版讲究一个兼容性，Linux 发行版打包主要讲究[可重复性](https://reproducible-builds.org/)和动态链接（能用系统库的就不要静态链接）。

大不了说 Linux 发行版打包的版本，你们不对其稳定性、安全性作出任何承诺，但客观上这个渠道是不可能作为上游能够堵住的（否则构成对使用、传播的限制，就不是开源软件了）。

longdafeng commented:
@yetist @xen0n 
谢谢你的建议, 在这个问题上, 你们也投入了很多的时间.  这是一个很大的特性. 

在流程上:
1. 一个大的feature 构思之前, 大家最好在github 上提一个discussion, 然后讨论这个feature 是不是要做, 然后怎么做. 如果要做转化为issue, 跟踪起来. 
2. 社区做事情, 往往讨论会比较多, 执行上不一定那么简单高效. 
3. 不要一上来, 就提一个巨大的pull request, 通常一个没有事前沟通的大的pull request 在很多开源社区容易被拒(在apache开源社区也是这样)

回到 LoongArch 支持问题, 我看我们的研发已经在和你们一起弄了一个分支. 我们先看能不能跑通了, 跑通后, 再看怎么集成到master上,  并且最小改动到主干, 可以尝试提供用户在 LoongArch 自助编译和打包 即可, 等充分测试后, 官方再发官方release 包

yetist commented:
> @yetist @xen0n 谢谢你的建议, 在这个问题上, 你们也投入了很多的时间. 这是一个很大的特性.

这是个大特性没错，对于大特性的支持，也是需要通过拆分成一个一个可构建、可验证的小补丁方式合入的，这个思路是没错的。

> 在流程上:
> 
> 1. 一个大的feature 构思之前, 大家最好在github 上提一个discussion, 然后讨论这个feature 是不是要做, 然后怎么做. 如果要做转化为issue, 跟踪起来.

之所以创建这个 PR，是因为有些问题表现出来是个综合性问题，不是简单通过issue讨论、或者创建一个单一PR就可以解决的，**需要通过可验证的代码进行呈现并加以讨论的**（如上面说的Debug可以编译，但Release却链接出错），所以希望在有可验证的代码的基础上**加以讨论**，这是创建本 PR 的初衷。

> 2. 社区做事情, 往往讨论会比较多, 执行上不一定那么简单高效.

可以看到，本 PR 一开始就加了 `WIP:`，原意是边讨论边修改验证，并非一上来就想直接合并。并且之前也说了，会拆分PR并独立提交，直到本PR的目标完成(如果社区认同此目标的话)。

> 3. 不要一上来, 就提一个巨大的pull request, 通常一个没有事前沟通的大的pull request 在很多开源社区容易被拒(在apache开源社区也是这样)

并非一上来就创建了一个巨大的 PR。比如像已经创建的这几个PR，都是独立只完成一件事情，并且不影响现有构建和功能：#832、 #846、#848

再比如本 PR 中的其他提交，如对bison高版本的支持、构建时自动生成 ob_error.{h,cpp}、使用pkgconfig检测依赖库、一些 bug fix等等，也会逐一拆分成独立补丁发起PR。

> 回到 LoongArch 支持问题, 我看我们的研发已经在和你们一起弄了一个分支. 我们先看能不能跑通了, 跑通后, 再看怎么集成到master上, 并且最小改动到主干, 可以尝试提供用户在 LoongArch 自助编译和打包 即可, 等充分测试后, 官方再发官方release 包

可以看到，此 PR 中并没有 LoongArch 相关的任何代码，虽然是在做 LoongArch 的过程中发现的问题，但是对于 LoongArch 的支持创建了独立且干净的 PR #851，并且我们内部先行做了 review https://github.com/loongarch64/oceanbase/pull/1 。

所以，创建这个PR，只是纯粹希望 OceanBase 可以做的更好，可能成为一个架构无关、发行版无关、对工具和库版本无特定要求的通用项目。

另外，再次强调，此 PR 带上 WIP  就是希望大家可以参与讨论，并且能就技术本身进行讨论，但是直到目前为止，并没有看到任何有关的技术讨论：https://github.com/oceanbase/oceanbase/pull/847#issuecomment-1080543731

yetist committed:
add build helper script

- use clang as default compiler
- add depends check
- use split build directory for gcc/clang

yetist committed:
create ld symlink to ld.lld for gcc

yetist commented:
> @whhe @LINxiansheng
> 
> 使用 `cmake -DCMAKE_BUILD_TYPE=Debug ...` 可正常编译出 `observer`；而使用`cmake -DCMAKE_BUILD_TYPE=Release ...` 编译时，在最后链接时会报错如下：
> 
> ```
> [100%] Linking CXX executable observer                                           
> cd /home/yetist/checkout/04-loongarch/oceanbase/build/src/observer && /usr/bin/cmake -E cmake_link_script CMakeFiles/observer.dir/link.txt --verbose=1
> /usr/bin/g++ -fdebug-prefix-map=/home/yetist/checkout/04-loongarch/oceanbase=. -fdiagnostics-color -O3 -DNDEBUG -Wl,--build-id=uuid -z noexecstack CMakeFiles/ob_main.dir/main.cpp.o -o observer  -Wl,-rpath,/home/yetist/checkout/04-loongarch/oceanbase/build/src/observer:/home/yetist/checkout/04-loongarch/oceanbase/build/deps/oblib/src:/home/yetist/checkout/04-loongarch/oceanbase/build/deps/easy/src liboceanbase_static.so -lgcc -lstdc++ ../../deps/oblib/src/lib/libmalloc_hook.a -L/home/yetist/checkout/04-loongarch/oceanbase/deps/3rd/usr/local/oceanbase/deps/devel/lib/mariadb -lmariadb -laio ../../deps/oblib/src/liboblib.so ../../deps/easy/src/libeasy.so -lisal -lunwind -ldl -lrt -lcrypto -lcurl -L/home/yetist/checkout/04-loongarch/oceanbase/deps/3rd/usr/local/oceanbase/deps/devel/lib -L/usr/lib -lssl -lcrypto -lpthread -lm 
> /usr/bin/ld: liboceanbase_static.so: undefined reference to `oceanbase::sql::ObQueryRange::get_param_value(oceanbase::common::ObObj&, oceanbase::common::Ob2DArray<oceanbase::common::ObObjParam, 2096128, oceanbase::common::ObWrapperAllocator, false, oceanbase::common::ObSEArray<oceanbase::common::ObObjParam*, 64l, oceanbase::common::ObWrapperAllocator, false> > const&) const'
> /usr/bin/ld: liboceanbase_static.so: undefined reference to `oceanbase::obrpc::ObLogicPartitionServiceRpcP<(oceanbase::obrpc::ObRpcPacketCode)1134>::~ObLogicPartitionServiceRpcP()'
> /usr/bin/ld: liboceanbase_static.so: undefined reference to `oceanbase::obrpc::ObLogicPartitionServiceRpcP<(oceanbase::obrpc::ObRpcPacketCode)1135>::~ObLogicPartitionServiceRpcP()'
> collect2: 错误：ld 返回 1                                                        
> make[2]: *** [src/observer/CMakeFiles/observer.dir/build.make:90：src/observer/observer] 错误 1
> ```
> 
> 请帮忙看看是代码存在问题还是工程存在问题？谢谢

更新一下状态，如果使用gcc+ld的话，就是上述错误。如果使用gcc+ld.lld的话，错误类似，详情如下：
```
[100%] Linking CXX executable observer
cd /media/DATA/archlinux-loongarch/oceanbase/build.gcc.release/src/observer && /usr/bin/cmake -E cmake_link_script CMakeFiles/observer.dir/link.txt --verbose=1
/usr/bin/g++ -fdebug-prefix-map=/media/DATA/archlinux-loongarch/oceanbase=. -fdiagnostics-color -B/media/DATA/archlinux-loongarch/oceanbase/deps/3rd/compile -O3 -DNDEBUG -Wl,--build-id=uuid -z noexecstack CMakeFiles/ob_main.dir/main.cpp.o -o observer  -Wl,--start-group liboceanbase_static.a -lgcc -lstdc++ ../../deps/oblib/src/lib/libmalloc_hook.a -L/media/DATA/archlinux-loongarch/oceanbase/deps/3rd/usr/local/oceanbase/deps/devel/lib/mariadb -lmariadb -laio ../../deps/oblib/src/liboblib.a ../../deps/easy/src/libeasy.a -Wl,--end-group -static-libgcc -static-libstdc++ -L/usr/lib -lssl -lcrypto -lm -L/media/DATA/archlinux-loongarch/oceanbase/deps/3rd/usr/local/oceanbase/deps/devel/lib -lisal -lunwind -lpthread -ldl -lrt -lcrypto -lcurl 
ld: error: undefined symbol: oceanbase::sql::ObQueryRange::get_param_value(oceanbase::common::ObObj&, oceanbase::common::Ob2DArray<oceanbase::common::ObObjParam, 2096128, oceanbase::common::ObWrapperAllocator, false, oceanbase::common::ObSEArray<oceanbase::common::ObObjParam*, 64l, oceanbase::common::ObWrapperAllocator, false> > const&) const
>>> referenced by ob_table_location.cpp
>>>               ob_table_location.cpp.o:(oceanbase::sql::ObTableLocation::calculate_partition_ids_fast_for_non_insert(oceanbase::sql::ObExecContext&, oceanbase::common::Ob2DArray<oceanbase::common::ObObjParam, 2096128, oceanbase::common::ObWrapperAllocator, false, oceanbase::common::ObSEArray<oceanbase::common::ObObjParam*, 64l, oceanbase::common::ObWrapperAllocator, false> > const&, oceanbase::common::ObObj&, bool&) const) in archive liboceanbase_static.a

ld: error: undefined symbol: oceanbase::obrpc::ObLogicPartitionServiceRpcP<(oceanbase::obrpc::ObRpcPacketCode)1135>::~ObLogicPartitionServiceRpcP()
>>> referenced by ob_srv_xlator_partition.cpp
>>>               ob_srv_xlator_partition.cpp.o:(oceanbase::observer::init_srv_xlator_for_migrator(oceanbase::observer::ObSrvRpcXlator*)::'lambda6'(oceanbase::observer::ObGlobalContext const&, oceanbase::rpc::frame::ObReqProcessor*&, oceanbase::obrpc::ObRpcSessionHandler&)::_FUN(oceanbase::observer::ObGlobalContext const&, oceanbase::rpc::frame::ObReqProcessor*&, oceanbase::obrpc::ObRpcSessionHandler&)) in archive liboceanbase_static.a

ld: error: undefined symbol: oceanbase::obrpc::ObLogicPartitionServiceRpcP<(oceanbase::obrpc::ObRpcPacketCode)1134>::~ObLogicPartitionServiceRpcP()
>>> referenced by ob_srv_xlator_partition.cpp
>>>               ob_srv_xlator_partition.cpp.o:(oceanbase::observer::init_srv_xlator_for_migrator(oceanbase::observer::ObSrvRpcXlator*)::'lambda5'(oceanbase::observer::ObGlobalContext const&, oceanbase::rpc::frame::ObReqProcessor*&, oceanbase::obrpc::ObRpcSessionHandler&)::_FUN(oceanbase::observer::ObGlobalContext const&, oceanbase::rpc::frame::ObReqProcessor*&, oceanbase::obrpc::ObRpcSessionHandler&)) in archive liboceanbase_static.a
collect2: 错误：ld 返回 1
make[2]: *** [src/observer/CMakeFiles/observer.dir/build.make:90：src/observer/observer] 错误 1
make[2]: 离开目录“/media/DATA/archlinux-loongarch/oceanbase/build.gcc.release”
make[1]: *** [CMakeFiles/Makefile2:10909：src/observer/CMakeFiles/observer.dir/all] 错误 2
make[1]: 离开目录“/media/DATA/archlinux-loongarch/oceanbase/build.gcc.release”
make: *** [Makefile:169：all] 错误 2
```

如果使用clang+lld 可以正常编译出来。

所以目前并不清楚是gcc的问题，还是代码需要修改。

注：以上编译测试环境为x86 archlinux，理论上x86 linux环境应该都能做验证。

yetist commented:
我是希望这个分支，至少在x86上可以做到 gcc/clang、release/debug的4种组合都能正常编译出来，然后再拆分补丁，分别创建不同的PR进行合并，希望社区有人能帮忙研究一下这个问题。

yetist commented:
本分支在 x86 linux 上最新编译测试情况如下：

toolchain   | debug | release |                                                  
----------- |-------|---------|                                                  
clang + ld  | ✅    |   ✅    |                                                  
clang + lld | ✅    |   ✅    |                                                  
gcc   + ld  | ✅    |   ❌    |                                                  
gcc   + lld | ✅    |   ❌    |
